pyiceberg/io/pyarrow.py

```diff
@@ -1304,6 +1305,195 @@ def _read_all_delete_files(fs: FileSystem, tasks: Iterable[FileScanTask]) -> Dic
+        """
+        return property_as_bool(self._io.properties, PYARROW_USE_LARGE_TYPES_ON_READ, True)
+
+    @property
+    def _projected_field_ids(self) -> Set[int]:
+        """Set of field IDs that should be projected from the data files."""
+        return {
+            id
+            for id in self._projected_schema.field_ids
+            if not isinstance(self._projected_schema.find_type(id), (MapType, ListType))
+        }.union(extract_field_ids(self._bound_row_filter))
+
+    def project_table(self, tasks: Iterable[FileScanTask]) -> pa.Table:
+        """Project the Iceberg table to a pa.Table.
+
+        Returns a pa.Table with data from the Iceberg table by resolving the
+        right columns that match the current table schema. Only data that
+        matches the provided row_filter expression is returned.
+
+        Args:
+            tasks: FileScanTasks representing the data files and delete files to read from.
+
+        Returns:
+            A PyArrow table. Result is capped at the limit, if specified.
+
+        Raises:
+            ResolveError: When a required field cannot be found in the file
+            ValueError: When a field type in the file cannot be projected to the schema type
+        """
+        deletes_per_file = _read_all_delete_files(self._fs, tasks)
+        executor = ExecutorFactory.get_or_create()
+
+        def _project_table_from_scan_task(task: FileScanTask) -> pa.Table:
+            batches = list(self._project_batches_from_scan_tasks_and_deletes([task], deletes_per_file))
+            if len(batches) > 0:
+                return pa.Table.from_batches(batches)
+            else:
+                return None
+
+        futures = [
+            executor.submit(
+                _project_table_from_scan_task,
+                task,
+            )
+            for task in tasks
+        ]
+        total_row_count = 0
+        # for consistent ordering, we need to maintain future order
+        futures_index = {f: i for i, f in enumerate(futures)}
+        completed_futures: SortedList[Future[pa.Table]] = SortedList(iterable=[], key=lambda f: futures_index[f])
+        for future in concurrent.futures.as_completed(futures):
+            completed_futures.add(future)
+            if table_result := future.result():
+                total_row_count += len(table_result)
+            # stop early if limit is satisfied
+            if self._limit is not None and total_row_count >= self._limit:
+                break
+
+        # by now, we've either completed all tasks or satisfied the limit
+        if self._limit is not None:
+            _ = [f.cancel() for f in futures if not f.done()]
+
+        tables = [f.result() for f in completed_futures if f.result()]
+
+        if len(tables) < 1:
+            return pa.Table.from_batches([], schema=schema_to_pyarrow(self._projected_schema, include_field_ids=False))
+
+        result = pa.concat_tables(tables, promote_options="permissive")
```

Comment:
this is something im curious about, and probably need to look into how PyArrow handles its internal data structure. 

Right now, `_project_table_from_scan_task` returns a `pa.Table` object by transforming batches to table
```
            batches = list(self._project_batches_from_scan_tasks_and_deletes([task], deletes_per_file))
            if len(batches) > 0:
                return pa.Table.from_batches(batches)
```
We then parallelize this and end up with multiple `pa.Table`s before joining all the tables together
```
        result = pa.concat_tables(tables, promote_options="permissive")
```
So two operations here
* record_batches to table
* joining tables into a single table

Is it more performant to return record batches from `_project_table_from_scan_task` and then join the record batches to the table?

Reply:
[pyarrow.concat_tables](https://arrow.apache.org/docs/python/generated/pyarrow.concat_tables.html) docs
[pyarrow.Table.from_batches](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.from_batches) docs

Not sure if these operations are 0 copy. 

@Fokko would you know about the performance characteristics of these two functions?

Reply:
for `concat_table`

> If promote_options=”none”, a zero-copy concatenation will be performed. The schemas of all the Tables must be the same (except the metadata), otherwise an exception will be raised. The result Table will share the metadata with the first table.
...
If promote_options=”permissive”, the behavior of default plus types will be promoted to the common denominator that fits all the fields.

Reply:
actually, not sure how much this matters performance-wise

Reply:
it should not be a lot of overhead to create tables from record batches; pyarrow does this in places internally already, iirc.  Table columns can be chunked so there's no copying required.

Reply:
Thank you for chiming in with your arrow expertise @corleyma . Always appreciative of your input here. That's consistent with my understanding as as well. Unless we have to cast the data to different types, the operations should be zero copy.

For the record, the refactoring should not introduce any behavior change from the current behavior, and here's the rationale for why we are doing the things currently the way they are:
- ThreadPoolExecutor converts record batches to tables, and then those tables are concatenated into a single table: This is to parallelize the work of eagerly materializing the data in the parquet files. If we return iterators of record batches and then create a pa.Table from the lazy iterator of record batches, we don't benefit from the parallelization.
- when using pa.Table, we can use the 'permissive' mode on concatenation. This means types need to be upcast only if there are discrepancies between the tables (like large vs small types). When creating a pa.Table from RecordBatches, the schema being read from different files need to be the same, which can be a challenge with small vs large types: https://arrow.apache.org/docs/python/generated/pyarrow.Table.html.

Given above reasons, we are materializing recordbatches as pa.Table  in individual threads that corresponds to the work of scanning individual data files, and then concatenating the tables into one.

Reply:
that makes sense, thanks for the explanation! lets stick with the current form